internal/builder/images.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2024 Stefan Prodan.
+// SPDX-License-Identifier: AGPL-3.0
+
+package builder
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/fluxcd/pkg/apis/kustomize"
+	ssautil "github.com/fluxcd/pkg/ssa/utils"
+	gcname "github.com/google/go-containerregistry/pkg/name"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/yaml"
+)
+
+// ComponentImage represents a container image used by a component.
+type ComponentImage struct {
+	Name       string
+	Repository string
+	Tag        string
+	Digest     string
+}
+
+// ExtractComponentImages reads the source directory and extracts the container images
+// from the components manifests.
+func ExtractComponentImages(srcDir string, opts Options) ([]ComponentImage, error) {
+	images := make([]ComponentImage, len(opts.Components))
+	for i, component := range opts.Components {
+		d, err := os.ReadFile(filepath.Join(srcDir, fmt.Sprintf("/%s.yaml", component)))
+		if err != nil {
+			return nil, err
+		}
+		objects, err := ssautil.ReadObjects(bytes.NewReader(d))
+		if err != nil {
+			return nil, err
+		}
+		for _, obj := range objects {
+			if obj.GetKind() == "Deployment" {
+				containers, ok, _ := unstructured.NestedSlice(obj.Object, "spec", "template", "spec", "containers")
+				if !ok {
+					return nil, fmt.Errorf("containers not found in %s", obj.GetName())
+				}
+				for _, container := range containers {
+					img := container.(map[string]interface{})["image"].(string)
+					tag, err := gcname.NewTag(img, gcname.WeakValidation)
+					if err != nil {
+						return nil, err
+					}
+
+					images[i] = ComponentImage{
+						Name:       component,
+						Repository: fmt.Sprintf("%s/%s", strings.TrimSuffix(opts.Registry, "/"), component),
+						Tag:        tag.Identifier(),
+					}
+				}
+			}
+		}
+	}
+
+	return images, nil
+}
+
+// FetchComponentImages fetches the components images from the distribution repository.
+func FetchComponentImages(opts Options) (images []ComponentImage, err error) {
+	registry := strings.TrimSuffix(opts.Registry, "/")
+	var distro string
+
+	switch registry {
+	case "fluxcd":
+		distro = "upstream-alpine"
+	case "ghcr.io/fluxcd":
+		distro = "upstream-alpine"
+	case "ghcr.io/controlplaneio-fluxcd/alpine":
+		distro = "enterprise-alpine"
+	case "ghcr.io/controlplaneio-fluxcd/distroless":
+		distro = "enterprise-distroless"
+	default:
+		return nil, fmt.Errorf("unsupported registry: %s", registry)
+	}
+
+	const ghRepo = "https://raw.githubusercontent.com/controlplaneio-fluxcd/distribution/main/images"
```

Comment:
maybe this should be a parameter? we could then fetch from a test http server in our tests? otherwise tests will fail everytime change the digests upstream.

Reply:
The tests will not fail, as they point to the upstream CNCF Flux images (fixed version v2.3.0). I'm reluctant on exposing the URL in the API, as this function should fetch the digests from the OCI repo directly, instead of using Git like now.